lib/gc/Analysis/MatmulConfigAnalysis.cpp

```diff
@@ -352,102 +384,110 @@ bool readConfigFromAttrs(MatmulConfig &config, ArrayRef<NamedAttribute> attrs) {
-    bool hasPredefinedConfig = readConfigFromAttrs(config, attrs);
+      // try to read the config from the attributes
+      SmallVector<NamedAttribute> attrs(linalgOp->getAttrs());
+      bool hasPredefinedConfig = readConfigFromAttrs(config, attrs);
 
-    // if there is a given config, skip the cost model
-    if (!hasPredefinedConfig) {
-      LLVM_DEBUG(llvm::dbgs() << "No predefined config\n");
-      // TODO: Could add a weight or priority for cost model
-      SmallVector<std::tuple<CostModelFn, std::string, double>> costModelList =
-          {{workloadBalancedCost, "workloadBalancedCost", 1},
-           {vectorRegEfficiencyCost, "vectorRegEfficiencyCost ", -1},
-           {computationIntensityOnL2Cache, "computationIntensityOnL2Cache", -1},
-           {memoryConsumptionOnThreadCost, "memoryConsumptionOnThreadCost",
-            -1}};
-      SmallVector<uint32_t> shape = {M, N, K};
-      std::vector<MatmulConfig> configCandidates =
-          prepareConfigCandidates(root, sysDesc, shape, givenInnermostBlock);
-      for (auto &&[fn, name, threshold] : costModelList)
-        configCandidates = filterConfigByCostModel(
-            configCandidates, linalgOp, shape, sysDesc, fn, 0.5, threshold);
-      if (!configCandidates.empty())
-        config = configCandidates[0];
-    }
+      // if there is a given config, skip the cost model
+      if (!hasPredefinedConfig) {
+        LLVM_DEBUG(llvm::dbgs() << "No predefined config\n");
+        // TODO: Could add a weight or priority for cost model
+        SmallVector<std::tuple<CostModelFn, std::string, double>>
+            costModelList = {
+                {dynamicBufferizationCost, "dynamicBufferizationCost", 0.9},
+                {workloadBalancedCost, "workloadBalancedCost", 1},
+                {vectorRegEfficiencyCost, "vectorRegEfficiencyCost ", -1},
+                {computationIntensityOnL2Cache, "computationIntensityOnL2Cache",
+                 -1},
+                {memoryConsumptionOnThreadCost, "memoryConsumptionOnThreadCost",
+                 -1}};
+        SmallVector<uint32_t> shape = {M, N, K};
+        std::vector<MatmulConfig> configCandidates =
+            prepareConfigCandidates(root, sysDesc, shape, givenInnermostBlock,
+                                    allowIndivisibleInnerBlock);
+        for (auto &&[fn, name, threshold] : costModelList)
+          configCandidates = filterConfigByCostModel(
+              configCandidates, linalgOp, shape, sysDesc, fn, 0.5, threshold);
+        if (!configCandidates.empty())
+          config = configCandidates[0];
+      }
 
-    LLVM_DEBUG(llvm::dbgs()
-               << "Final config\nNumThreads: " << sysDesc.getNumThreads()
-               << ", MatmulConfig: " << config << "\n");
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Final config\nNumThreads: " << sysDesc.getNumThreads()
+                 << ", MatmulConfig: " << config << "\n");
+    }
+    hasConfig = true;
   }
+  return config;
```

Comment:
It might be better to validate the `config` before returning? for both calculated and user-provided?

Reply:
Thanks for the suggestion! I reuse the `validateConfig` here